src/mrack/providers/beaker.py

```diff
@@ -0,0 +1,334 @@
+
+        job_id = self.hub.jobs.upload(job.toxml())  # schedule beaker job
+
+        return (job_id, req["name"])
+
+    def _get_host_info_from_prov_result(self, prov_result):
+        """Get needed host infromation from Beaker provisioning result."""
+        # init the dict
+        try:
+            ip_address = socket.gethostbyname(prov_result["system"])
+        except socket.gaierror:
+            ip_address = None
+
+        result = {
+            "id": prov_result["JobID"],
+            "name": prov_result["hname"],
+            "addresses": [ip_address],
+            "status": prov_result["status"],
+            "fault": prov_result["result"] if prov_result["result"] != "Pass" else None,
+        }
+
+        return result
+
+    def _get_recipe_info(self, beaker_id):
+        bkr_job_xml = self.hub.taskactions.to_xml(beaker_id).encode("utf8")
+
+        resources = []
+        for recipe in eTree.fromstring(bkr_job_xml).iter("recipe"):
+            resources.append(
+                {
+                    "family": recipe.get("family"),
+                    "distro": recipe.get("distro"),
+                    "arch": recipe.get("arch"),
+                    "variant": recipe.get("variant"),
+                    "system": recipe.get("system"),
+                    "status": recipe.get("status"),
+                    "result": recipe.get("result"),
+                    "rid": recipe.get("id"),
+                    "id": recipe.get("job_id"),
+                }
+            )
+
+        return resources[0] if len(resources) == 1 else []
+
+    async def wait_till_provisioned(
+        self, bkr_id_name, timeout=None, poll_sleep=None, max_attempts=None
+    ):
+        """Wait for Beaker provisioning result."""
+        beaker_id, req_hname = bkr_id_name
+
+        if not poll_sleep:
+            poll_sleep = self.poll_sleep
+        if not timeout:
+            timeout = self.timeout
+        if not max_attempts:
+            max_attempts = self.max_attempts
+
+        resource = {}
+        attempts = 0
+        prev_status = ""
+
+        while attempts < self.max_attempts:
+            attempts += 1
+            resource = self._get_recipe_info(beaker_id)
+            status = resource["status"]
+
+            if prev_status != status:
+                logger.info(
+                    f"Job {beaker_id} has changed status ("
+                    f"{prev_status} -> {status})"
+                )
+                prev_status = status
+
+            if STATUS_MAP.get(status) == STATUS_ACTIVE:
+                break
+            elif STATUS_MAP.get(status) == STATUS_PROVISIONING:
+                await asyncio.sleep(self.poll_sleep)
+            elif STATUS_MAP.get(status) in [STATUS_ERROR, STATUS_DELETED]:
+                logger.warning(
+                    f"Job {beaker_id} has errored with status "
+                    f"{status} with result {resource['result']}"
+                )
```

Comment:
Shouldn't here be also break? And probably in the case below as well. Otherwise it is iterating without a sleep.

Reply:
My bad :)